adoc/chapters/programming_interface.adoc

```diff
@@ -21621,1365 +20463,5193 @@ multi-dimensional groups, the order of work-items in group [code]#g# is
+It stores the integral part to the object pointed to by [code]#iptr#.
+
+_Returns:_ The fractional part of the argument [code]#x#.
+
+*Overload (4):*
+
+_Constraints:_ Available only if all of the following conditions are met:
+
+* [code]#NonScalar# is [code]#marray#, [code]#vec#, or the
+  [code]#+__swizzled_vec__+# type with element type [code]#float#,
+  [code]#double#, or [code]#half#;
+* [code]#Ptr# is [code]#multi_ptr# with [code]#ElementType# equal to
+  [code]#NonScalar#, unless [code]#NonScalar# is the [code]#+__swizzled_vec__+#
+  type, in which case the [code]#ElementType# is the corresponding
+  [code]#vec#; and
+* [code]#Ptr# is [code]#multi_ptr# with [code]#Space# equal to one of the
+  _writeable address spaces_ as defined above.
+
+_Effects:_ The [code]#modf# function breaks each element of the argument
+[code]#x# into integral and fractional parts, each of which has the same sign
+as the element.  It stores the integral parts of each element to the object
+pointed to by [code]#iptr#.
+
+_Returns:_ The fractional parts of each element of the argument [code]#x#.
+
+The return type is [code]#NonScalar# unless [code]#NonScalar# is the
+[code]#+__swizzled_vec__+# type, in which case the return type is the
+corresponding [code]#vec#.
 
 a@
+[frame=all,grid=none]
+!====
+a!
 [source]
 ----
-vfloatn nan(vuint32n nancode)
-mfloatn nan(muintn nancode)
-float nan(unsigned int nancode)
-vdoublen nan(vuint64n nancode)
-mdoublen nan(mulongn nancode)
-double nan(unsigned long nancode)
-vhalfn nan(vuint16n nancode)
-mhalfn nan(mushortn nancode)
-half nan(unsigned short nancode)
+float nan(unsigned int nancode)         (1)
+double nan(unsigned long nancode)       (2)
+half nan(unsigned short nancode)        (3)
+
+template<typename NonScalar>            (4)
+/*return-type*/ nan(NonScalar nancode)
 ----
-   a@ Returns a quiet NaN. The nancode may be placed
-      in the significand of the resulting NaN.
+!====
+
+*Overloads (1) - (3):*
+
+_Returns:_ A quiet NaN. The [code]#nancode# may be placed
+in the significand of the resulting NaN.
+
+*Overload (4):*
+
+_Constraints:_ Available only if one of the following conditions is met:
+
+* [code]#NonScalar# is [code]#marray# and the element type is
+  [code]#unsigned int#, [code]#unsigned long#, or [code]#unsigned short#; or
+* [code]#NonScalar# is [code]#vec# or the [code]#+__swizzled_vec__+# type and
+  the element type is [code]#uint32_t#, [code]#uint64_t#, or [code]#uint16_t#.
+
+_Returns:_ A quiet NaN for each element of [code]#nancode#.  Each
+[code]#nancode[i]# may be placed in the significand of the resulting NaN.
+
+The return type depends on [code]#NonScalar#:
+[options="header"]
+!====
+!NonScalar ! Return Type
+![code]#marray<unsigned int, N># ![code]#marray<float, N>#
+![code]#marray<unsigned long, N># ![code]#marray<double, N>#
+![code]#marray<unsigned short, N># ![code]#marray<half, N>#
+
+![code]#vec<uint32_t, N># +
+ [code]#+__swizzled_vec__<uint32_t, N>+#
```

Comment:
We could just consider `__swizzled_vec__` as a magic type, including the power to handle template parameters as a magical detail. Otherwise we might have some similar issues somewhere else too?

Reply:
I scanned through the SYCL spec, and I did not see any other cases where we assume `__swizzled_vec__` has a defined set of template parameters.